File: core/src/commonMain/kotlin/com/encatch/core/InlineSlotRegistry.kt
package com.encatch.core

/**
 * Inline slot registry, mirroring `form-presentation-registry.ts`.
 *
 * Maintains an ordered list of mounted inline slots (e.g. `EncatchInlineFormView`s in the
 * `:android` module). When `showForm` fires, [resolvePresentationTarget] determines whether
 * it should render inline (matching slot found) or fall through to the modal form dialog.
 *
 * Routing rules:
 *  1. Exact match — first slot whose formId matches the payload ids wins.
 *  2. Wildcard    — first slot with no formId catches anything not exact-matched.
 *  3. Modal       — no inline slot registered or none match.
 */
data class InlineSlot(val slotId: String, val formId: String? = null)

sealed class PresentationTarget {
    data class Inline(val slotId: String) : PresentationTarget()
    object Modal : PresentationTarget()
}

object InlineSlotRegistry {
    private val slots = mutableListOf<InlineSlot>()

    /**
     * Registers a new inline slot on view attach. Returns an opaque slotId to use with
     * [unregisterInlineSlot] / [updateInlineSlot]. Registration order is preserved —
     * first-registered wins for wildcard resolution.
     */
    fun registerInlineSlot(formId: String? = null): String {
        val slotId = uuidV7()
        slots.add(InlineSlot(slotId, formId))
        return slotId
    }

    /** Removes an inline slot on view detach. */
    fun unregisterInlineSlot(slotId: String) {
        slots.removeAll { it.slotId == slotId }
    }

    /** Updates the formId of an existing slot without changing its registration order. */
    fun updateInlineSlot(slotId: String, formId: String?) {
        val index = slots.indexOfFirst { it.slotId == slotId }
        if (index != -1) slots[index] = slots[index].copy(formId = formId)
    }

    /**
     * Determines whether the given showForm payload should render inline or modal.
     *
     * ID matching checks the slot's formId against [formId] (the slug/uuid passed by the
     * caller, or formConfigurationId) and [feedbackConfigurationId] (server-resolved id).
     * Single pass: first exact match wins, then the first wildcard; else modal.
     */
    fun resolvePresentationTarget(formId: String, feedbackConfigurationId: String?): PresentationTarget {
        val candidateIds = setOfNotNull(formId.takeIf { it.isNotEmpty() }, feedbackConfigurationId?.takeIf { it.isNotEmpty() })

        var firstWildcard: InlineSlot? = null
        for (slot in slots) {
            if (slot.formId != null) {
                if (slot.formId in candidateIds) return PresentationTarget.Inline(slot.slotId)
            } else if (firstWildcard == null) {
                firstWildcard = slot
            }
        }

        return firstWildcard?.let { PresentationTarget.Inline(it.slotId) } ?: PresentationTarget.Modal
    }

    /** Exposed for testing only — do not use in production code. */
    fun slotsSnapshot(): List<InlineSlot> = slots.toList()

    /** Exposed for testing only — clears the registry. */
    fun clearSlots() {
        slots.clear()
    }
}

File: swift/Sources/Encatch/Core/InlineSlotRegistry.swift
import Foundation

/// Inline slot registry, mirroring `form-presentation-registry.ts`.
///
/// Maintains an ordered list of mounted inline slots (e.g. `EncatchInlineFormView`s). When
/// `showForm` fires, `resolvePresentationTarget` determines whether it should render inline
/// (matching slot found) or fall through to the modal form dialog.
///
/// Routing rules:
///  1. Exact match — first slot whose formId matches the payload ids wins.
///  2. Wildcard    — first slot with no formId catches anything not exact-matched.
///  3. Modal       — no inline slot registered or none match.
public struct InlineSlot: Sendable, Equatable {
    public var slotId: String
    public var formId: String?

    public init(slotId: String, formId: String? = nil) {
        self.slotId = slotId
        self.formId = formId
    }
}

public enum PresentationTarget: Sendable, Equatable {
    case inline(slotId: String)
    case modal
}

/// Class (not a plain namespace) since it holds mutable state — one shared instance, mirrors the
/// Kotlin `object`.
public final class InlineSlotRegistry: @unchecked Sendable {
    public static let shared = InlineSlotRegistry()

    private let lock = NSLock()
    private var slots: [InlineSlot] = []

    private init() {}

    /// Registers a new inline slot on view attach. Returns an opaque slotId to use with
    /// `unregisterInlineSlot`/`updateInlineSlot`. Registration order is preserved — first-registered
    /// wins for wildcard resolution.
    public func registerInlineSlot(formId: String? = nil) -> String {
        let slotId = uuidV7()
        lock.lock()
        slots.append(InlineSlot(slotId: slotId, formId: formId))
        lock.unlock()
        return slotId
    }

    /// Removes an inline slot on view detach.
    public func unregisterInlineSlot(_ slotId: String) {
        lock.lock()
        slots.removeAll { $0.slotId == slotId }
        lock.unlock()
    }

    /// Updates the formId of an existing slot without changing its registration order.
    public func updateInlineSlot(_ slotId: String, formId: String?) {
        lock.lock()
        if let index = slots.firstIndex(where: { $0.slotId == slotId }) {
            slots[index].formId = formId
        }
        lock.unlock()
    }

    /// Determines whether the given showForm payload should render inline or modal.
    ///
    /// ID matching checks the slot's formId against `formId` (the slug/uuid passed by the
    /// caller, or formConfigurationId) and `feedbackConfigurationId` (server-resolved id).
    /// Single pass: first exact match wins, then the first wildcard; else modal.
    public func resolvePresentationTarget(formId: String, feedbackConfigurationId: String?) -> PresentationTarget {
        var candidateIds = Set<String>()
        if !formId.isEmpty { candidateIds.insert(formId) }
        if let feedbackConfigurationId, !feedbackConfigurationId.isEmpty {
            candidateIds.insert(feedbackConfigurationId)
        }

        lock.lock()
        let snapshot = slots
        lock.unlock()

        var firstWildcard: InlineSlot?
        for slot in snapshot {
            if let slotFormId = slot.formId {
                if candidateIds.contains(slotFormId) {
                    return .inline(slotId: slot.slotId)
                }
            } else if firstWildcard == nil {
                firstWildcard = slot
            }
        }

        if let firstWildcard {
            return .inline(slotId: firstWildcard.slotId)
        }
        return .modal
    }

    /// Exposed for testing only — do not use in production code.
    public func slotsSnapshot() -> [InlineSlot] {
        lock.lock()
        defer { lock.unlock() }
        return slots
    }

    /// Exposed for testing only — clears the registry.
    public func clearSlots() {
        lock.lock()
        slots.removeAll()
        lock.unlock()
    }
}
